exo-setup/src/app-setup.spec.ls

```diff
@@ -0,0 +1,102 @@
+require! {
+  '../../exosphere-shared': {run-process}
+  'chai': {expect}
+  'fs-extra' : fs
+  'js-yaml' : yaml
+  'mkdirp'
+  'path'
+  'prelude-ls': {each}
+  'rimraf'
+}
+
+
+describe 'AppSetup', ->
+
+  describe 'set up a complex app' (...) ->
+    @timeout 600000
+
+    before (done) ~>
+      @app-name = 'complex-setup-app'
+      @app-dir = path.join process.cwd!, 'tmp', @app-name
+      @docker-compose-location = path.join @app-dir, 'tmp', 'docker-compose.yml'
+      @exocom-name = 'exocom0.22.1'
+      @internal-services = ['html-server', 'todo-service']
+      @external-services = ['external-service']
+      @external-dependencies = ['mongo3.4.0']
+      @process = checkout-setup-app @app-name, @app-dir
+        ..on 'ended' done
+
+    specify 'should create docker-compose.yml at the expected location', (done) ~>
+      fs.stat @docker-compose-location, (err, stat) ~>
+        expect(err).to.be.null
+        @docker-compose = yaml.safe-load fs.read-file-sync(@docker-compose-location, 'utf8')
+        done!
+
+    specify 'should list all services and dependencies under \'services\'', (done) ~>
+      all-services = [@exocom-name] ++ @internal-services ++ @external-dependencies ++ @external-services
+      expect(Object.keys @docker-compose.services).to.have.members(all-services)
+      done!
+
+    specify 'should generate an image name for each dependency and external service', (done) ~>
+      expect(@docker-compose.services[@exocom-name].image).to.not.be.empty
+      @external-dependencies.for-each (dependency) ~> expect(@docker-compose.services[dependency].image).to.not.be.empty
+      @external-services.for-each (service) ~> expect(@docker-compose.services[service].image).to.not.be.empty
+      done!
+
+    specify 'should generate a container name for each service and dependency', (done) ~>
+      expect(@docker-compose.services[@exocom-name].container_name).to.not.be.empty
+      @external-dependencies.for-each (dependency) ~> expect(@docker-compose.services[dependency].container_name).to.not.be.empty
+      @external-services.for-each (service) ~> expect(@docker-compose.services[service].container_name).to.not.be.empty
+      @internal-services.for-each (service) ~> expect(@docker-compose.services[service].container_name).to.not.be.empty
+      done!
+
+    specify 'should have the correct build command for each service and dependency', (done) ~>
+      @internal-services.for-each (service) ~> expect(@docker-compose.services[service].command).to.eql('echo "does not run"')
+      expect(@docker-compose.services[@exocom-name].command).to.eql('bin/exocom')
+      done!
+
+    specify 'should have the correct build path for each internal service', (done) ~>
+      @internal-services.for-each (service) ~> expect(@docker-compose.services[service].build).to.eql("../#service")
```

Comment:
Hm, I just ran it through the lsc compiler too, didn't know you could do this. Good to know these tests weren't returning a false positive though.